File: imagesExtractSave.js
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import fetch from 'node-fetch';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// 递归获取所有md文件
function findMarkdownFiles(dir) {
    let results = [];
    const files = fs.readdirSync(dir);
    
    for(const file of files) {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);
        
        if(stat.isDirectory() && !filePath.includes('node_modules')) {
            results = results.concat(findMarkdownFiles(filePath));
        } else if(path.extname(file) === '.md') {
            results.push(filePath);
        }
    }
    
    return results;
}

// 从markdown提取图片链接
function extractImageUrls(content) {
    const regex = /!\[.*?\]\((.*?)\)/g;
    const urls = [];
    let match;
    
    while((match = regex.exec(content)) !== null) {
        urls.push(match[1]);
    }
    
    return urls;
}

// 生成文件名
function generateFileName(url) {
    // 从 URL 中提取最后一部分作为文件名
    const urlParts = url.split('/');
    const fileName = urlParts[urlParts.length - 1];
    
    // 如果无法获取文件名，则使用之前的哈希方法
    if (!fileName || fileName.trim() === '') {
        const ext = path.extname(url) || '.png';
        const hash = crypto.createHash('md5').update(url).digest('hex');
        return hash.substring(0, 8) + ext;
    }
    
    return fileName;
}

// 下载图片
async function downloadImage(url, fileName, outputDir = 'images') {
    try {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch ${url}: ${response.statusText}`);
        }
        const buffer = await response.buffer();
        const imagePath = path.join(outputDir, fileName);
        
        // 确保输出目录存在
        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }
        
        fs.writeFileSync(imagePath, buffer);
        console.log(`Downloaded: ${fileName}`);
    } catch (error) {
        console.error(`Error downloading ${url}:`, error.message);
    }
}

// 主函数
export async function main(options = {}) {
    const {
        inputDir = 'docs',
        outputDir = 'images',
        mapFile = 'image-map.txt'
    } = options;

    const mdFiles = findMarkdownFiles(inputDir);
    const imageMap = new Map();
    
    for(const file of mdFiles) {
        const content = fs.readFileSync(file, 'utf-8');
        const urls = extractImageUrls(content);
        
        for(const url of urls) {
            if(url.startsWith('http')) {
                const fileName = generateFileName(url);
                imageMap.set(url, fileName);
            }
        }
    }
    
    // 保存映射关系
    const output = Array.from(imageMap.entries())
        .map(([url, fileName]) => `${url}\t${fileName}`)
        .join('\n');
        
    fs.writeFileSync(mapFile, output);
    
    console.log(`找到 ${imageMap.size} 个图片链接`);
    console.log(`映射关系已保存到 ${mapFile}`);
    
    // 下载所有图片
    console.log('开始下载图片...');
    const downloads = Array.from(imageMap.entries()).map(([url, fileName]) => 
        downloadImage(url, fileName, outputDir)
    );
    
    await Promise.all(downloads);
    console.log('所有图片下载完成！');
}

// 如果直接运行此文件
if (process.argv[1] === fileURLToPath(import.meta.url)) {
    main().catch(console.error);
}

File: imagesExtractSave.py
import os
import hashlib
import requests
from pathlib import Path

def find_markdown_files(directory):
    """递归获取所有md文件"""
    results = []
    for root, dirs, files in os.walk(directory):
        if 'node_modules' in root:
            continue
        for file in files:
            if file.endswith('.md'):
                results.append(os.path.join(root, file))
    return results

def extract_image_urls(content):
    """从markdown提取图片链接"""
    import re
    pattern = r'!\[.*?\]\((.*?)\)'
    return re.findall(pattern, content)

def generate_filename(url):
    """生成文件名"""
    # 从 URL 中提取最后一部分作为文件名
    filename = url.split('/')[-1]
    
    # 如果无法获取文件名，则使用哈希方法
    if not filename or filename.strip() == '':
        ext = os.path.splitext(url)[1] or '.png'
        hash_value = hashlib.md5(url.encode()).hexdigest()
        return hash_value[:8] + ext
    
    return filename

def download_image(url, filename):
    """下载图片"""
    try:
        response = requests.get(url)
        response.raise_for_status()
        
        # 确保 images 目录存在
        images_dir = Path('images')
        images_dir.mkdir(exist_ok=True)
        
        image_path = images_dir / filename
        with open(image_path, 'wb') as f:
            f.write(response.content)
        print(f'Downloaded: {filename}')
    except Exception as error:
        print(f'Error downloading {url}: {str(error)}')

async def main():
    """主函数"""
    md_files = find_markdown_files('docs')
    image_map = {}
    
    for file in md_files:
        with open(file, 'r', encoding='utf-8') as f:
            content = f.read()
            urls = extract_image_urls(content)
            
            for url in urls:
                if url.startswith('http'):
                    filename = generate_filename(url)
                    image_map[url] = filename
    
    # 保存映射关系
    with open('image-map.txt', 'w', encoding='utf-8') as f:
        f.write('\n'.join(f'{url}\t{filename}' for url, filename in image_map.items()))
    
    print(f'找到 {len(image_map)} 个图片链接')
    print('映射关系已保存到 image-map.txt')
    
    # 下载所有图片
    print('开始下载图片...')
    import asyncio
    import aiohttp
    
    async def download_with_aiohttp(url, filename):
        async with aiohttp.ClientSession() as session:
            try:
                async with session.get(url) as response:
                    if response.status == 200:
                        content = await response.read()
                        images_dir = Path('images')
                        images_dir.mkdir(exist_ok=True)
                        image_path = images_dir / filename
                        with open(image_path, 'wb') as f:
                            f.write(content)
                        print(f'Downloaded: {filename}')
                    else:
                        print(f'Failed to download {url}: {response.status}')
            except Exception as error:
                print(f'Error downloading {url}: {str(error)}')
    
    tasks = [download_with_aiohttp(url, filename) for url, filename in image_map.items()]
    await asyncio.gather(*tasks)
    print('所有图片下载完成！')

if __name__ == '__main__':
    import asyncio
    asyncio.run(main()) 